src/Microsoft.ApplicationInsights/Extensibility/Implementation/CorrelationLookup/ApplicationInsightsCorrelationIdProvider.cs

```diff
@@ -0,0 +1,130 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.CorrelationLookup
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading.Tasks;
+    using Extensibility;
+
+    /// <summary>
+    /// This <see cref="ApplicationInsightsCorrelationIdProvider"/> will query the Application Insights' Breeze endpoint to lookup an application id based on Instrumentation Key.
+    /// This will cache lookup results to prevent repeat queries.
+    /// This will rely on the <see cref="ProfileServiceWrapper" /> and <see cref="FailedRequestsManager" /> to record failed requests and block additional failing requests.
+    /// </summary>
+    public sealed class ApplicationInsightsCorrelationIdProvider : ICorrelationIdProvider, IDisposable
+    {
+        internal ConcurrentDictionary<string, bool> FetchTasks = new ConcurrentDictionary<string, bool>();
+
+        /// <summary>
+        /// Max number of Application Ids to cache.
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        private readonly ProfileServiceWrapper applicationIdProvider;
+
+        private ConcurrentDictionary<string, string> knownCorrelationIds = new ConcurrentDictionary<string, string>();
+
+        /// <summary>
+        /// Initialize a new instance of the <see cref="ApplicationInsightsCorrelationIdProvider"/> class.
+        /// </summary>
+        public ApplicationInsightsCorrelationIdProvider()
+        {
+            this.applicationIdProvider = new ProfileServiceWrapper();
+        }
+
+        /// <summary>
+        /// Unit Test Only! Initializes a new instance of the <see cref="ApplicationInsightsCorrelationIdProvider" /> class and accepts mocks for fetching Application Id.
+        /// </summary>
+        internal ApplicationInsightsCorrelationIdProvider(ProfileServiceWrapper profileServiceWrapper)
+        {
+            this.applicationIdProvider = profileServiceWrapper;
+        }
+
+        /// <summary>
+        /// Gets or sets the endpoint that is to be used to get the Application Insights resource's profile (Application Id etc.). 
+        /// Default value is "https://dc.services.visualstudio.com/api/profiles/{0}/appId". If this is overwritten, MUST include the '{0}' for string replacement!
```

Comment:
@SergeyKanzhelev How do you feel about this?
The alternative would be to ask users to overwrite the base "https://dc.services.visualstudio.com" and then we append the relative URI ourselves.

Second question: if we keep this should we add an EventSource log if a user fails to include the string replacement token?

Reply:
I think override is a rare scenario. If user sophisticated enough to do it - replacement token should be specified. I wonder would named replacement tokens work here? Like https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated So it's clear that {0} is instrumentation key. Not sure though how it might work...

Reply:
Interpolation wouldn't work here because the string needs a reference to a specific object. It gets compiled as a `String.Format()`. (Note that in those documented examples the object already exists where the string is defined).

I'll leave it as is and make sure it's well documented. thanks!